src/libopenrave/libopenrave.h

```diff
@@ -481,6 +481,152 @@ void UpdateOrCreateInfo(const rapidjson::Value& value, std::vector<boost::shared
+    // Note that we allocate full strings rather than string views because otherwise deserializing into the existing info struct may mutate the view out from under the map
+    std::unordered_map<std::string, size_t> infoIdxByName;
+    std::unordered_map<std::string, size_t> infoIdxById;
+    infoIdxByName.reserve(vInfosOut.size());
+    infoIdxById.reserve(vInfosOut.size());
+    for (size_t i = 0; i < vInfosOut.size(); i++) {
+        const boost::shared_ptr<T>& info = vInfosOut[i];
+        // Note that we emplace here rather than overwriting - we want to prioritize elements that area earlier in the vInfosOut array in the case of a duplicate
+        // This replicates the old behaviour, which involved a sequential array scan.
+        if (!info->_name.empty()) {
+            infoIdxByName.emplace(info->_name, i);
+        }
+        if (!info->_id.empty()) {
+            infoIdxById.emplace(info->_id, i);
+        }
+    }
+
+    // Track whether we deleted any elements and need to garbage collect the output vector before returning
+    bool didEraseElements = false;
+    for (rapidjson::Value::ConstValueIterator it = sourceItBegin; it != sourceItEnd; it++) {
+        // Bind the next JSON value to consider
+        const rapidjson::Value& value = *it;
+
+        // If the ID is non-empty, attempt to look up a matching entry in vInfosOut
+        const std::string id = OpenRAVE::orjson::GetStringJsonValueByKey(value, "id");
+        const std::string name = OpenRAVE::orjson::GetStringJsonValueByKey(value, pNameInJson);
+        OpenRAVE::orjson::GetCStringJsonValueByKey(value, pNameInJson);
+        size_t existingInfoIndex = -1;
+        if (!id.empty()) {
+            decltype(infoIdxById)::iterator existingInfoIdxIt = infoIdxById.find(id);
+            if (existingInfoIdxIt != infoIdxById.end()) {
+                existingInfoIndex = existingInfoIdxIt->second;
+            }
+        }
+        else {
+            // Check if the new info has a name. If it doesn't, don't try and match it - just always create a new object
+            // sometimes names can be empty, in which case, always create a new object
+            if (!name.empty()) {
+                decltype(infoIdxByName)::iterator existingInfoIdxIt = infoIdxByName.find(name);
+                if (existingInfoIdxIt != infoIdxByName.end()) {
+                    existingInfoIndex = existingInfoIdxIt->second;
+                }
+            }
+        }
+
+        // If this element is deleted, we need to remove any matching element from the info list
+        bool isDeleted = OpenRAVE::orjson::GetJsonValueByKey<bool>(value, "__deleted__", false);
+        if (existingInfoIndex != -1) {
+            if (isDeleted) {
+                // Erase the referents to this element from our skip lists, but only if they actually point to this index
+                decltype(infoIdxById)::iterator idIt = infoIdxById.find(vInfosOut[existingInfoIndex]->_id);
```

Comment:
this `find` and checks in the next line can be avoided if you reuse the iterator `existingInfoIdxIt` created the above block.

Reply:
These are intentionally different - the first check is using the name/id of the input JSON, and the second uses the name/id of the matched deserialized object. The JSON may only specify _one_ of name/id, but the deserialized object may have the other already loaded. Comment added to explain.